### DISCUSSION TITLE
Getting and Updating Listings

### DISCUSSION THREAD

**Discussion by JesusR25** [amzn/selling-partner-api-models#4346] [Q&A]
Hello, I am looking for a way to retrieve the entire listing of my account. The listings endpoint didn't work for me, so I used the /reports endpoint to get the GET_MERCHANT_LISTINGS_ALL_DATA report.

Now, what I want to do is keep this listing updated in my database through notifications.

When saving the listing obtained from the report, I find two possible approaches:

Retrieve all the SKUs from my report and make a query to the /listing endpoint to fetch and save that information in my DB. 

Save the report data exactly as it comes, without making additional queries.

I'm not sure which of the two approaches to take, as I assume I need to consider the structure of the notification. Could you help recommend which approach to follow? Or if you can help me resolve the error?

Thank you so much!

**Comment by ritikacd** [comment]
Hi @JesusR25 

We recommend the following approach:
1. One time call to GET_MERCHANT_LISTINGS_ALL_DATA reports to get all the listings, active and inactive. You can see the list of  report attributes [here](https://developer-docs.amazon.com/sp-api/docs/report-type-values-inventory#all-listings-report). 
2. Make a request to [getListingsItem](https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#getlistingsitem) operation to get the summaries and issues, if present for the listing. See the [includedData](https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#includeddata-subgroup-1) to see the datasets present. 
3. Rely on [LISTINGS_ITEM_STATUS_CHANGE](https://developer-docs.amazon.com/sp-api/docs/notification-type-values#listings_item_status_change) and  [LISTINGS_ITEM_ISSUES_CHANGE](https://developer-docs.amazon.com/sp-api/docs/notification-type-values#listings_item_issues_change)  notifications to get the updates and issues for the listings, only make subsequent call to getListingItem if you need to retrieve more details about the issues. 

For more information on error handling, refer to the [Handling listings item issues ](https://developer-docs.amazon.com/sp-api/docs/building-listings-management-workflows-guide#handling-listings-item-issues) guide. 


Thanks,
Ritika

**Reply by ajunisouk** [reply]
Hi @ritikacd, I was trying to list the products via sp api. Products got published but I haven't got any `LISTINGS_ITEM_STATUS_CHANGE` notification even though the product becomes DISCOVERABLE/BUYABLE. I was getting ISSUE_CHANGE, MFN_QUANTITY change notifications.

**Reply by ritikacd** [reply]
Hi @ajunisouk 

If you have no issues with your listing submissions, and it changed the status. Then you should receive LISTINGS_ITEM_STATUS_CHANGE notification. 

To investigate the issue further, please [open a support case](https://developer.amazonservices.com/support) following the guidance [here](https://developer.amazonservices.com/support-guidelines)https://developer.amazonservices.com/support-guidelines).

Thanks,
Ritika
Selling Partner API Services Team

**Reply by ajunisouk** [reply]
Hi @ritikacd, the product become buyable still we haven't got any event related to LISTINGS_ITEM_STATUS_CHANGE, while we got other notifications like ISSUE_CHANGE, MFN_QUANTITY. (We will raise a support ticket) Thanks!
Few other questions, if you know about these.
1. Ritika, is it possible to handle the entire product upload/listing process programmatically using the SP-APIs? We are facing many challenges and bugs, and the documentation isn't very clear.  
2. Does Amazon provide any on-call support for queries related to the Amazon SP-APIs?  
3. Is there any company that has implemented a complete programmatic solution using the Amazon SP-APIs?

**Reply by ritikacd** [reply]
> Hi @ritikacd, the product become buyable still we haven't got any event related to LISTINGS_ITEM_STATUS_CHANGE, while we got other notifications like ISSUE_CHANGE, MFN_QUANTITY. (We will raise a support ticket) Thanks! Few other questions, if you know about these.
> 
> 1. Ritika, is it possible to handle the entire product upload/listing process programmatically using the SP-APIs? We are facing many challenges and bugs, and the documentation isn't very clear.
> 2. Does Amazon provide any on-call support for queries related to the Amazon SP-APIs?
> 3. Is there any company that has implemented a complete programmatic solution using the Amazon SP-APIs?

Hi @ajunisouk,
1. [JSON_LISTINGS_FEED](https://developer-docs.amazon.com/sp-api/docs/listings-feed-type-values#listings-feed) feed type supports creating/uploading upto 10,000 items at a time. Refer to the [Building Listings Management Workflows Guide](https://developer-docs.amazon.com/sp-api/docs/building-listings-management-workflows-guide) as it covers all the commonly used listings workflows and best practices.
2. You can currently use GitHub Issues or discussions section to ask questions about the APIs or developer services for troubleshoot problems. If you contact the support team, be sure to include details such as the request, response, and timestamp to help them investigate the issue more efficiently.
3. You can view the list of developers who have implemented Listings solutions on the [Selling Partner Appstore](https://sellercentral.amazon.com/selling-partner-appstore/search?category=2ebe9108-1e62-11e8-b467-0ed5f89f718b&marketplacesIds=US&pageNumber=1&pageSize=48&locale=en-US). 

I hope this helps! 

Thank you,
Ritika
Selling Partner API Services

**Reply by abhaysharma0901** [reply]
Hi @ritikacd,  
I have a query about the syncing of the listings for an Amazon seller.

Currently, the process works like this:  
1. I first fetch the **"GET_MERCHANT_LISTINGS_ALL_DATA"** report to get a list of all ASINs.
2. Then, I parse those ASINs from the report and proceed to use **searchCatalogItem** API to fetch additional details, specifically the **mainImageUrl** and **productType**, because these two values are not included in the original report.

However, this process is quite slow. For example, if a seller has 6000 listings, this results in around 300 separate API requests (since each request can handle 20 ASINs). The issue is that I'm particularly after the **productType** information, which is necessary to update inventory using the **patchListingItem** API.

So, I'm wondering if there’s any way to optimize this process and reduce the number of API calls. Ideally, I want to get all the required details for all listings in just **4-5 API calls** rather than hundreds.

Can you suggest a more efficient way to achieve this?

**Reply by jawheral** [reply]
Hi @abhaysharma0901,

If you just need to patch inventory or prices using the `patchListingsItem` operation, you can just use the parent product type `PRODUCT`. There is no need to fetch the product type for each listing.

Best,
Jawher
Selling Partner Developer Services
